icu4c/source/common/unistr.cpp

```diff
@@ -1453,6 +1453,15 @@ UnicodeString::doReplace(int32_t start,
     srcLength = u_strlen(srcChars + srcStart);
   }
 
+  // Check for insertion into ourself
+  if (isBufferWritable()
+      && getArrayStart() < srcChars + srcStart + srcLength
```

Comment:
Consider moving up the oldArray declaration, initializing it with getArrayStart(), and using oldArray instead of multiple getArrayStart() calls.

Below, where oldArray comes from, you would use that instead of fUnion.fStackFields.fBuffer but still reset it, and remove the "else". (GitHub won't let me add comments there.)